File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch04/Python/P04-SortQuart.py
# Sorting the elements of an array into quartiles
from sort import *
from random import *

# main

n = 12                                        # number of values to be sorted
x = [0]*(n+1)                                            # array to be sorted
ind = [0]*(n+1); rnk = [0]*(n+1)                # arrays of indexes and ranks

print("Original array:")
for i in range(1,n+1):
   x[i] = random()                                # random sub-unitary values
   print("{0:6.2f}".format(x[i]),end="")
print("\n")

Index(x,ind,n)                                      # create array of indexes
Rank(ind,rnk,n)                                       # create array of ranks

print("Quartile 1")
print("     i   ind   rnk        x")
for i in range(1,int(n/4)+1):
   indi = ind[i]
   print("{0:6d}{1:6d}{2:6d}{3:10.2f}".format(i,ind[i],rnk[indi],x[indi]))

print("Quartile 2")
print("     i   ind   rnk        x")
for i in range(int(n/4)+1,int(n/2)+1):
   indi = ind[i]
   print("{0:6d}{1:6d}{2:6d}{3:10.2f}".format(i,ind[i],rnk[indi],x[indi]))

print("Quartile 3")
print("     i   ind   rnk        x")
for i in range(int(n/2)+1,int(3*n/4)+1):
   indi = ind[i]
   print("{0:6d}{1:6d}{2:6d}{3:10.2f}".format(i,ind[i],rnk[indi],x[indi]))

print("Quartile 4")
print("     i   ind   rnk        x")
for i in range(int(3*n/4)+1,n+1):
   indi = ind[i]
   print("{0:6d}{1:6d}{2:6d}{3:10.2f}".format(i,ind[i],rnk[indi],x[indi]))

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch04/C/P04-SortQuart.cpp
// Sorting the elements of an array into quartiles
#include <stdio.h>
#include <stdlib.h>
#include "sort.h"

int main()
{
   const int n = 12;                         // number of values to be sorted
   double x[n+1];                                       // array to be sorted
   int ind[n+1], rnk[n+1];                     // arrays of indexes and ranks
   int i, indi;

   printf("Original array:\n");
   for (i=1; i<=n; i++) {
      x[i] = rand()/(RAND_MAX+1e0);              // random sub-unitary values
      printf("%6.2f",x[i]);
   }
   printf("\n");

   Index(x,ind,n);                                 // create array of indexes
   Rank(ind,rnk,n);                                  // create array of ranks

   printf("Quartile 1\n");
   printf("     i   ind   rnk        x\n");
   for (i=1; i<=int(n/4); i++) {
      indi = ind[i];
      printf("%6i%6i%6i%10.2f\n",i,ind[i],rnk[indi],x[indi]);
   }

   printf("Quartile 2\n");
   printf("     i   ind   rnk        x\n");
   for (i=int(n/4)+1; i<=int(n/2); i++) {
      indi = ind[i];
      printf("%6i%6i%6i%10.2f\n",i,ind[i],rnk[indi],x[indi]);
   }

   printf("Quartile 3\n");
   printf("     i   ind   rnk        x\n");
   for (i=int(n/2)+1; i<=int(3*n/4); i++) {
      indi = ind[i];
      printf("%6i%6i%6i%10.2f\n",i,ind[i],rnk[indi],x[indi]);
   }

   printf("Quartile 4\n");
   printf("     i   ind   rnk        x\n");
   for (i=int(3*n/4)+1; i<=n; i++) {
      indi = ind[i];
      printf("%6i%6i%6i%10.2f\n",i,ind[i],rnk[indi],x[indi]);
   }
}
